resin/resin/dialog/file_dialog.hpp

```diff
@@ -0,0 +1,62 @@
+#ifndef RESIN_FILE_DIALOG_HPP
+#define RESIN_FILE_DIALOG_HPP
+
+#include <future>
+#include <libresin/utils/logger.hpp>
+#include <nfd/nfd.hpp>
+#include <optional>
+#include <string>
+#include <thread>
+#include <vector>
+
+namespace resin {
+
+class FileDialog {
+ public:
+  FileDialog(FileDialog&&)            = delete;
+  FileDialog(FileDialog&)             = delete;
+  FileDialog& operator=(FileDialog&)  = delete;
+  FileDialog& operator=(FileDialog&&) = delete;
+
+  ~FileDialog();
+
+  static FileDialog& instance() {
+    static FileDialog instance;
+    return instance;
+  }
+
+  inline bool is_active() const { return dialog_task_.has_value(); }
+
+  inline void open(
+      const std::function<void(const std::filesystem::path&)>& on_open_function,
+      std::optional<std::vector<std::pair<const std::string_view, const std::string_view>>> filters = std::nullopt) {
+    on_finish_ = on_open_function;
+    start_file_dialog(true, std::move(filters));
+  }
+
+  inline void save(
+      const std::function<void(const std::filesystem::path&)>& on_save_function,
+      std::optional<std::vector<std::pair<const std::string_view, const std::string_view>>> filters = std::nullopt,
+      std::optional<std::string> default_name                                                       = std::nullopt) {
+    on_finish_ = on_save_function;
+    start_file_dialog(false, std::move(filters), std::move(default_name));
+  }
+
+  void update();
+
+ private:
+  FileDialog() = default;
+  void start_file_dialog(
+      bool is_open,
+      std::optional<std::vector<std::pair<const std::string_view, const std::string_view>>> filters = std::nullopt,
```

Comment:
I don't think we utilise the fact that `filters` is a std::vector anywhere, other than that it is iterable. Could we switch the type to std::span?

Reply:
note that when using spans we need to ensure that the viewed array is alive through the whole thread delegate lifetime, which means that when calling it in e.g. sdf tree component we would need to use static for the array definition

Reply:
which might be what we want to achieve considering a rather static nature of the content

Reply:
Yes, I think it should be the programmers responsibility to ensure that filters are alive for the lifetime of the dialog thread. Additionally, in most scenarios filters are constant. If the programmer wants to dynamically generate a list of filters (for some reason) then they should assure, that the filters don't die before the dialog thread completes its job